src/draive/openai/chat.py

```diff
@@ -46,34 +61,58 @@ async def openai_chat_completion(
     input: StringConvertible,  # noqa: A002
     history: list[ConversationMessage] | None = None,
     toolset: Toolset | None = None,
-    stream: bool = False,
-) -> OpenAIChatStream | str:
+    stream: StreamingProgressUpdate[OpenAIChatStreamingPart] | bool = False,
+) -> OpenAIChatStream | str | None:
     config: OpenAIChatConfig = ctx.state(OpenAIChatConfig)
     async with ctx.nested("openai_chat_completion"):
         client: OpenAIClient = ctx.dependency(OpenAIClient)
-        if stream:
-            return await _chat_stream(
-                client=client,
-                config=config,
-                messages=_prepare_messages(
-                    instruction=instruction,
-                    history=history or [],
-                    input=input,
-                ),
-                toolset=toolset,
-            )
-
-        else:
-            return await _chat_completion(
-                client=client,
-                config=config,
-                messages=_prepare_messages(
-                    instruction=instruction,
-                    history=history or [],
-                    input=input,
-                ),
-                toolset=toolset,
-            )
+        match stream:
+            case False:
+                return await _chat_response(
+                    client=client,
+                    config=config,
+                    messages=_prepare_messages(
+                        instruction=instruction,
+                        history=history or [],
+                        input=input,
+                    ),
+                    toolset=toolset,
+                )
+
+            case True:
+
+                async def stream_task(
```

Comment:
I'd make this one a private method, same as `_chat_stream`. Somehow I always dislike defining methods inside methods :/

Reply:
it is more than private 😉  and it captures the context - can't be defined outside